src/verbs/queue_pair.rs

```diff
@@ -38,6 +42,122 @@ pub enum QueuePairState {
     Unknown = ibv_qp_state::IBV_QPS_UNKNOWN,
 }
 
+impl From<u32> for QueuePairState {
+    fn from(state: u32) -> Self {
+        match state {
+            ibv_qp_state::IBV_QPS_RESET => QueuePairState::Reset,
+            ibv_qp_state::IBV_QPS_INIT => QueuePairState::Init,
+            ibv_qp_state::IBV_QPS_RTR => QueuePairState::ReadyToReceive,
+            ibv_qp_state::IBV_QPS_RTS => QueuePairState::ReadyToSend,
+            ibv_qp_state::IBV_QPS_SQD => QueuePairState::SendQueueDrain,
+            ibv_qp_state::IBV_QPS_SQE => QueuePairState::SendQueueError,
+            ibv_qp_state::IBV_QPS_ERR => QueuePairState::Error,
+            ibv_qp_state::IBV_QPS_UNKNOWN => QueuePairState::Unknown,
+            _ => panic!("Unknown qp state: {state}"),
+        }
+    }
+}
+
+// According to C standard, enums should be int, but Rust just uses whatever
+// type returned by Clang, which is uint on Linux platforms, so just cast it
+// into int.
+//
+// https://github.com/rust-lang/rust-bindgen/issues/1966
+//
+#[bitmask(i32)]
+#[bitmask_config(vec_debug)]
+pub enum QueuePairAttributeMask {
+    State = ibv_qp_attr_mask::IBV_QP_STATE.0 as _,
+    CurrentState = ibv_qp_attr_mask::IBV_QP_CUR_STATE.0 as _,
+    EnableSendQueueDrainedAyscNotify = ibv_qp_attr_mask::IBV_QP_EN_SQD_ASYNC_NOTIFY.0 as _,
+    AccessFlags = ibv_qp_attr_mask::IBV_QP_ACCESS_FLAGS.0 as _,
+    PartitionKeyIndex = ibv_qp_attr_mask::IBV_QP_PKEY_INDEX.0 as _,
+    Port = ibv_qp_attr_mask::IBV_QP_PORT.0 as _,
+    QueueKey = ibv_qp_attr_mask::IBV_QP_QKEY.0 as _,
+    AddressVector = ibv_qp_attr_mask::IBV_QP_AV.0 as _,
+    PathMtu = ibv_qp_attr_mask::IBV_QP_PATH_MTU.0 as _,
+    Timeout = ibv_qp_attr_mask::IBV_QP_TIMEOUT.0 as _,
+    RetryCount = ibv_qp_attr_mask::IBV_QP_RETRY_CNT.0 as _,
+    ResponderNotReadyRetryCount = ibv_qp_attr_mask::IBV_QP_RNR_RETRY.0 as _,
+    ReceiveQueuePacketSequenceNumber = ibv_qp_attr_mask::IBV_QP_RQ_PSN.0 as _,
+    MaxReadAtomic = ibv_qp_attr_mask::IBV_QP_MAX_QP_RD_ATOMIC.0 as _,
+    AlternativePath = ibv_qp_attr_mask::IBV_QP_ALT_PATH.0 as _,
+    MinResponderNotReadyTimer = ibv_qp_attr_mask::IBV_QP_MIN_RNR_TIMER.0 as _,
+    SendQueuePacketSequenceNumber = ibv_qp_attr_mask::IBV_QP_SQ_PSN.0 as _,
+    MaxDestinationReadAtomic = ibv_qp_attr_mask::IBV_QP_MAX_DEST_RD_ATOMIC.0 as _,
+    PathMigrationState = ibv_qp_attr_mask::IBV_QP_PATH_MIG_STATE.0 as _,
+    Capabilities = ibv_qp_attr_mask::IBV_QP_CAP.0 as _,
+    DestinationQueuePairNumber = ibv_qp_attr_mask::IBV_QP_DEST_QPN.0 as _,
+    RateLimit = ibv_qp_attr_mask::IBV_QP_RATE_LIMIT.0 as _,
+}
+
+// Define the vital and optional mask according to the spec, so that we
+// could provide attr check for users. Furthermore, provide more useful
+// error messages.
+const RcQpRequiredMaskForAnyToReset: QueuePairAttributeMask = QueuePairAttributeMask::State;
```

Comment:
These `const` variables are a little bit ugly, though I don't figure out a way to handle this better.
I can't put them in the `QueuePairAttributeMask` definition as when we need to debug invalid and needed masks, all of these possible combinations would be printed.